docs/gui/journal.rst

```diff
@@ -0,0 +1,45 @@
+gui/journal
+===========
+
+.. dfhack-tool::
+    :summary: Fort journal with a multi-line text editor.
+    :tags: fort interface
+
+The `gui/journal` plugin provides a handy tool to document
+the important details for the fortresses.
+
+With this multi-line rich text editor,
+you can keep track of your fortress's background story, goals, notable events,
+and both short-term and long-term plans.
+
+This is particularly useful when you need to take a longer break from the game.
+Having detailed notes makes it much easier to resume your game after
+a few weekds or months, without losing track of your progress and objectives.
+
+Supported Features
+------------------
+
+- Cursor Control: Navigate through text using arrow keys (left, right, up, down) for precise cursor placement.
+- Fast Rewind: Use 'Shift+Left'/'Ctrl+B' and 'Shift+Right'/'Ctrl+F'to move the cursor one word back or forward
+- Longest X Position Memory: The cursor remembers the longest x position when moving up or down, making vertical navigation more intuitive.
+- Mouse Control: Use the mouse to position the cursor within the text, providing an alternative to keyboard navigation.
+- New Lines: Easily insert new lines using the submit key, supporting multiline text input.
+- Text Wrapping: Text automatically wraps within the editor, ensuring lines fit within the display without manual adjustments.
+- Backspace Support: Use the backspace key to delete characters to the left of the cursor.
+- Delete Character: 'Ctrl+D' deletes the character under the cursor.
+- Line Navigation: 'Ctrl+H' (like "Home") moves the cursor to the beginning of the current line, and 'Ctrl+E' (like "End") moves it to the end.
+- Delete Current Line: 'Ctrl+U' deletes the entire current line where the cursor is located.
+- Delete Rest of Line: 'Ctrl+K' deletes text from the cursor to the end of the line.
+- Delete Last Word: 'Ctrl+W' removes the word immediately before the cursor.
+- Clipboard Operations: Perform local cut, copy, and paste operations on selected text or the current line using 'Ctrl+X', 'Ctrl+C', and 'Ctrl+V'.
```

Comment:
I tried it, but it looks like `dfhack.internal.getClipboardTextCp437` do not works properly with multiline text.

`dfhack.internal.setClipboardTextCp437Multiline` store the text in the clipboard correctly, but when restored, I get some `?` characters. Also, there are some other bugs - lines with this new `?` char can no longer be copied (`dfhack.internal.setClipboardTextCp437Multiline` do not throw, but there no text in memory when copied such line)

Reply:
it sounds like we need an explicit multiline clipboard get call. I'll see if I can add that for you

Reply:
added `dfhack.internal.getClipboardTextCp437Multiline()` in https://github.com/DFHack/dfhack/pull/4759

Reply:
I can not make it work.

both `dfhack.internal.getClipboardTextCp437Multiline` and `dfhack.internal.setClipboardTextCp437Multiline` seems to have no effect now in my tests.
but no-multiline version works fine for both funs.

I pulled newest `develop` and rebuilt `dfhack`.

Reply:
What does something like this look like for you?

https://github.com/DFHack/scripts/assets/977482/3fe93e04-fb5a-4209-9a3b-e6c4fad56ab6

Reply:
it works and return proper array. I did not expect the function return an array, as all other from this family works with strings.

also, it seems that `setClipboardTextCp437Multiline` works properly only when it as least single newline character in string. if there is no new lines in given string, `setClipboardTextCp437Multiline` do not work.
is it expected?

Reply:
> also, it seems that setClipboardTextCp437Multiline works properly only when it as least single newline character in string. if there is no new lines in given string, setClipboardTextCp437Multiline do not work.
is it expected?

I'll look into that.

regarding returning an array -- remember that there *is* no newline in CP437. If we treat a string as "multiline", it is an agreement for a particular API that it is giving the string non-standard treatment. That's why this function returns an array instead of embedding character 10 -- it's so you can differentiate between CP437 character 10 (◙) and a string that is intended to be on a separate line

Reply:
> setClipboardTextCp437Multiline works properly only when it as least single newline

fixed

Reply:
Thanks! I migrated journal to use OS clipboard, seems to works fine now.

Reply:
this doc line needs updating to match the current behavior. Also, it would be good to document what copy/paste will do in various situations.

For example, if nothing is selected, Ctrl-C will behave as if the entire current line is selected and copy the current line, including the terminating newline if there is a natural newline. if text is selected, only that text will be copied

Before documenting this behavior, though, maybe we should decide what is correct. In a text editor (which we are generally trying to emulate here) triple clicking will select the entire "natural" line, not just the wrapped portion that was under the mouse when it was triple clicked.


https://github.com/DFHack/scripts/assets/977482/2720e6ac-a207-4cdb-ad5c-ec4f356365bb

I think triple clicking, and copy (when no text is explicitly selected) should act on the "natural" line, not the wrapped line.

Reply:
I am agree and going to implement triple line fix and update docs soon.